contracts/AutomationBotAggregator.sol

```diff
@@ -0,0 +1,181 @@
+            serviceRegistry.getRegisteredService(AUTOMATION_BOT_KEY)
+        );
+        AutomationBotAggregator automationAggregatorBot = AutomationBotAggregator(
+            serviceRegistry.getRegisteredService(AUTOMATION_AGGREGATOR_BOT_KEY)
+        );
+        IValidator validator = IValidator(getValidatorAddress(groupTypeId));
+
+        require(validator.validate(replacedTriggerId, triggersData), "aggregator/validation-error");
+        (uint256[] memory cdpIds, uint256[] memory triggerTypes) = validator.decode(triggersData);
+
+        uint256 firstTriggerId = automationBot.triggersCounter() + 1;
+        uint256[] memory triggerIds = new uint256[](triggersData.length);
+        for (uint256 i = 0; i < triggerTypes.length; i++) {
+            (bool status, ) = address(automationBot).delegatecall(
+                abi.encodeWithSelector(
+                    AutomationBot(automationBot).addTrigger.selector,
+                    cdpIds[i],
+                    triggerTypes[i],
+                    replacedTriggerId[i],
+                    triggersData[i]
+                )
+            );
+
+            triggerIds[i] = firstTriggerId + i;
+            require(status, "aggregator/add-trigger-fail");
+        }
+
+        automationAggregatorBot.addRecord(cdpIds[0], groupTypeId, triggerIds);
+    }
+
+    function removeTriggerGroup(
+        uint256 cdpId,
+        uint256 groupId,
+        uint256[] memory triggerIds,
+        bool removeAllowance
+    ) external onlyDelegate {
+        AutomationBot automationBot = AutomationBot(
+            serviceRegistry.getRegisteredService(AUTOMATION_BOT_KEY)
+        );
+        AutomationBotAggregator automationAggregatorBot = AutomationBotAggregator(
+            serviceRegistry.getRegisteredService(AUTOMATION_AGGREGATOR_BOT_KEY)
+        );
+
+        for (uint256 i = 0; i < triggerIds.length; i++) {
+            (bool status, ) = address(automationBot).delegatecall(
+                abi.encodeWithSelector(
+                    AutomationBot(automationBot).removeTrigger.selector,
+                    cdpId,
+                    triggerIds[i],
+                    removeAllowance
+                )
+            );
+            require(status, "aggregator/remove-trigger-fail");
+        }
+
+        automationAggregatorBot.removeRecord(cdpId, groupId, triggerIds);
+    }
+
+    function addRecord(
+        uint256 cdpId,
+        uint16 groupTypeId,
+        uint256[] memory triggerIds
+    ) external {
+        ManagerLike manager = ManagerLike(serviceRegistry.getRegisteredService(CDP_MANAGER_KEY));
+
+        require(isCdpAllowed(cdpId, msg.sender, manager), "aggregator/no-permissions");
+        triggerGroupCounter++;
+        bytes32 hash = getTriggerGroupHash(triggerGroupCounter, cdpId, triggerIds);
+        activeTriggerGroups[triggerGroupCounter] = TriggerGroupRecord(hash, cdpId);
+
+        emit TriggerGroupAdded(triggerGroupCounter, groupTypeId, triggerIds, hash);
```

Comment:
ok, i take back my prev comment, hash isn't necessary here. but cdpId is definitely needed. otherwise we'll have a major headache indexing these events